Add support for function breakpoints

thegecko opened the pull request:
Signed-off-by: Rob Moran <rob.moran@arm.com>

Introduced support for adding function breakpoints, see:

https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_function-breakpoints

jonahgraham commented on src/mi/breakpoint.ts:
```diff
@@ -66,3 +66,8 @@ export function sendBreakDelete(gdb: GDBBackend, request: {
 }
+
+export function sendBreakFunctionInsert(gdb: GDBBackend, fn: string) {
+    const command = `-break-insert -t --function ${fn}`;
```
Intentional temporary breakpoint?

thegecko replied:
No, thanks

thegecko commented:
Tests would be a lot easier to write if they worked on MacOS or ran on PRs (with the output shown)

jonahgraham commented:
> Tests would be a lot easier to write if they worked on MacOS or ran on PRs (with the output shown)

Please add details on Issue #12 - or raise a new issue if needed.

thegecko commented:
> I think this still doesn't meet the DAP spec

Still WiP

thegecko commented:
Rebased and added tests, still need to resolve the issue with not removing them...

jonahgraham commented:
> @thegecko pushed 1 commit.
>
>    88fcd1f Support removing function breakpoints

@thegecko does this mean it is ready for review? Let me know when/if it is and I will try to get it in.

thegecko commented:
@jonahgraham This should be good to go. I've got a change which de-duplicates the breakpoint resolving in `setBreakpoints` and `setFunctionBreakpoints`, but felt that should be a PR after this one. Let me know if you'd prefer to see it added here, however.

jonahgraham commented:
> Let me know if you'd prefer to see it added here, however.

Lets avoid the duplication in the first place - if you are ready to add it now, please do that. It will save me having to compare the setBreakpoints and setFunctionBreakpoints versions.

thegecko commented:
@jonahgraham, I've introduced a common `resolveBreakpoints()` function for determining the breakpoints to add and delete across `setBreakpoints()` and `setFunctionBreakpoints()`.

I've introduced a few more tests to exercise this new common function and I've tested it all locally.

It's probably worth focussing your checks on the breakpoint handling overall as there is some churn in that area.

I've also added a few minor bug-fixes and comments.

jonahgraham commented on src/GDBDebugSession.ts:
```diff
@@ -328,20 +339,83 @@ export class GDBDebugSession extends LoggingDebugSession {
+                (vsbp, gdbbp) => {
+
+                // Always invalidate hit conditions as they have a one-way mapping to gdb ignore and temporary
+                if (vsbp.hitCondition) {
```
`hitCondition` and `condition` - VSCode [does not yet have](https://github.com/microsoft/vscode/issues/3646) UI for this so it seems acceptable to not implement support for now, but we are at some point going to need our own fork of the debug protocol spec to document our differences and extensions and this is one of them.

jonahgraham commented on src/integration-tests/functionBreakpoints.spec.ts:
```diff
@@ -0,0 +1,70 @@
+        const event = await dc.waitForEvent('stopped') as StoppedEvent;
+        expect(event.body.reason).to.eq('function breakpoint');
+        const scope = await getScopes(dc);
+        expect(scope.frame.line).to.eq(2);
```
There is already a helper that does this:

```
        dc.assertStoppedLocation('function breakpoint', {line: 2});
```
